### ISSUE TITLE
[ty] Unify staticness checks and avoid inspecting irrelevant generic declaration metadata

### ISSUE THREAD

**Issue by carljm** [astral-sh/ty#4238]
`Type::is_fully_static` (currently used in `unsound-return-type` checks) currently checks whether a recursive type visitor encounters any dynamic types:

```rust
pub(crate) fn is_fully_static(self, db: &'db dyn Db, env: &ProgramEnvironment) -> bool {
    dynamic_content(db, env, self).is_absent()
}
```

For nominal generic classes, this visitor traverses the generic declaration context as well as the actual specialization. As a result, an `Any` appearing in a type parameter’s bound or default can cause a fully concrete specialization to be wrongly classified as non-static.

For example:

```python
from typing import Any

class Plain[T]:
    item: T

class BoundedAny[T: Any]:
    item: T

class DefaultAny[T = Any]:
    item: T

def plain(value: Any) -> Plain[int]:
    return value

def bounded(value: Any) -> BoundedAny[int]:
    return value

def defaulted(value: Any) -> DefaultAny[int]:
    return value
```

With `unsound-return-statement` enabled, ty reports the expected diagnostic for `plain`, but not for `bounded` or `defaulted`. The declarations’ `Any` bound/default do not make `BoundedAny[int]` or `DefaultAny[int]` gradual: both types are specialized with `int`.

Their materializations confirm this:

```python
from ty_extensions import Bottom, Top

def materializations(
    bounded_top: Top[BoundedAny[int]],
    bounded_bottom: Bottom[BoundedAny[int]],
    defaulted_top: Top[DefaultAny[int]],
    defaulted_bottom: Bottom[DefaultAny[int]],
) -> None:
    reveal_type(bounded_top)       # BoundedAny[int]
    reveal_type(bounded_bottom)    # BoundedAny[int]
    reveal_type(defaulted_top)     # DefaultAny[int]
    reveal_type(defaulted_bottom)  # DefaultAny[int]
```

There are currently two inconsistent definitions of staticness in the implementation:

1. `Type::is_fully_static` uses `dynamic_content(...).is_absent()`.
2. Several constraint-solving paths classify a type as static when its top and bottom materializations are equal.

The materialization-based definition correctly ignores irrelevant declaration metadata in the examples above, but replacing `is_fully_static` with direct materialization equality is not currently safe:

- `TypedDictType::apply_type_mapping_impl` explicitly notes that materialization of gradual `TypedDict`s is incomplete. A class-based `TypedDict` containing an `Any` field may therefore retain identical top and bottom materializations even though it is not fully static.
- Class-backed protocol materialization preserves distinct top/bottom wrapper identities even when the protocol’s exposed requirements are fully static. Direct `Type` equality can therefore reject semantically static protocols.
- Recursive aliases, protocols, and `TypedDict`s require careful cycle handling; the existing visitor deliberately returns an indeterminate result for some growing recursive specializations.
- Materializations are Salsa-cached, so using both materializations everywhere could increase retained cache memory and should be benchmarked.

A reasonable first step would be to update the dynamic-content visitor so nominal generic aliases inspect their actual specialized arguments without treating declaration bounds/defaults as dynamic content. The protocol-specific visitor already applies this principle.

Longer term, we should determine whether a single semantic staticness predicate can replace both existing approaches. A materialization-based predicate becomes more attractive if gradual `TypedDict` materialization is completed, static protocol materializations are canonicalized or compared appropriately, recursive cases remain safe, and the performance/memory impact is acceptable.

Regression coverage should include:

- Generic classes specialized with static arguments despite `Any` bounds, constraints, or defaults.
- Generic classes whose actual specialized arguments are gradual.
- Static and gradual class-based protocols, including inherited members and properties.
- Static and gradual `TypedDict`s, including inherited fields and extra items.
- Recursive aliases, recursively specialized protocols, and recursive `TypedDict`s.
- Both `unsound-return-statement` and `unsound-yield`, which currently depend on `is_fully_static`.
- Existing constraint-solver paths that independently compare top and bottom materializations.
